src/derived.rs

```diff
@@ -740,34 +776,54 @@ where
             .is_some();
 
         // Either way, we have to update our entry.
+        //
+        // Keep in mind, though, we only acquired a read lock so a lot
+        // could have happened in the interim. =) Therefore, we have
+        // to probe the current state of `key` and in some cases we
+        // ought to do nothing.
         {
             let mut map = self.map.write();
-            if maybe_changed {
-                map.remove(key);
-            } else {
-                match map.get_mut(key) {
-                    Some(QueryState::Memoized(memo)) => {
-                        // It is possible that other threads were verifying inputs
-                        // at the same time. They too will be mutating the
-                        // map. However, they can only come to the same conclusion
-                        // that we did.
+            match map.get_mut(key) {
```

Comment:
I *think* that is correct, but perhaps we should block in the `InProgress` case?

Reply:
I think i'll leave it as is.